server/src/main/java/org/apache/iotdb/db/query/dataset/groupby/GroupByLevelDataSet.java

```diff
@@ -0,0 +1,217 @@
+      records.add(tmpRecord);
+      logger.info(tmpRecord.toString());
+    }
+  }
+
+  @Override
+  protected boolean hasNextWithoutConstraint() throws IOException {
+    return index < records.size();
+  }
+
+  @Override
+  protected RowRecord nextWithoutConstraint() {
+    return records.get(index++);
+  }
+
+  private void initGroupBy()
+    throws QueryProcessException, StorageEngineException {
+    // get all aggregation results, then we package them to one record
+    for (int i = 0; i < paths.size(); i++) {
+      Path path = paths.get(i);
+      if (!pathExecutors.containsKey(path)) {
+        //init GroupByExecutor
+        pathExecutors.put(path,
+          getGroupByExecutor(path, groupByPlan.getAllMeasurementsInDevice(path.getDevice()), dataTypes.get(i), this.context, null, null));
+        resultIndexes.put(path, new ArrayList<>());
+      } else {
+        throw new QueryProcessException("duplicated path found, path:" + path);
+      }
+      resultIndexes.get(path).add(i);
+      AggregateResult aggrResult = AggregateResultFactory
+        .getAggrResultByName(groupByPlan.getDeduplicatedAggregations().get(i), dataTypes.get(i));
+      pathExecutors.get(path).addAggregateResult(aggrResult);
+    }
+  }
+
+  private GroupByExecutor getGroupByExecutor(Path path, Set<String> allSensors, TSDataType dataType,
+                                             QueryContext context, Filter timeFilter, TsFileFilter fileFilter)
+    throws StorageEngineException, QueryProcessException {
+    return new LocalGroupByExecutor(path, allSensors, dataType, context, timeFilter, fileFilter);
+  }
+
+  private RowRecord getRecordWithoutTimeInterval()
+    throws IOException {
+    RowRecord record = new RowRecord(0);
+    AggregateResult[] fields = new AggregateResult[paths.size()];
+
+    try {
+      for (Map.Entry<Path, GroupByExecutor> pathToExecutorEntry : pathExecutors.entrySet()) {
+        GroupByExecutor executor = pathToExecutorEntry.getValue();
+        List<AggregateResult> aggregations = executor.calcResult(Long.MIN_VALUE, Long.MAX_VALUE);
+        for (int i = 0; i < aggregations.size(); i++) {
+          int resultIndex = resultIndexes.get(pathToExecutorEntry.getKey()).get(i);
+          fields[resultIndex] = aggregations.get(i);
+        }
+      }
+    } catch (QueryProcessException e) {
+      logger.error("GroupByWithoutValueFilterDataSet execute has error", e);
+      throw new IOException(e.getMessage(), e);
+    }
+
+    for (AggregateResult res : fields) {
+      if (res == null) {
+        record.addField(null);
+        continue;
+      }
+      record.addField(res.getResult(), res.getResultDataType());
+    }
+    return record;
+  }
+
+  private RowRecord mergeRecords(RowRecord newRecord, RowRecord oldRecord) {
+    if (oldRecord.getFields().isEmpty()) {
+      oldRecord.setFields(newRecord.getFields());
+    } else {
+      assert oldRecord.getFields().size() == newRecord.getFields().size();
```

Comment:
I think you'd better use `logger.error` instead of assertion, what do you think?

Reply:
ok, assert will help me found bug fastly. I can change it to error.

Reply:
And, I will maintain the User Guide docluments after merge.